driver: i2c: Add TCA9546a I2C switch driver

kurddt opened the pull request:
The chip allows to have up to 4 downlink channels.

The channels are exposed as devices so they can be used as standard i2c controllers

The driver only support use cases where the channels are used in mutual
exclusion.

kurddt commented:
Please let me know how to fix the CODEOWNER failure

galak commented:
> Please let me know how to fix the CODEOWNER failure

You need to add entries in CODEOWNER file for the new files that this PR adds.  Are you willing to maintain this driver (ie make fixes or changes as issues are raised)?

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,213 @@
+struct tca9546a_root_data {
+	struct k_mutex lock;
+	uint8_t select_reg;
+#ifde CONFIG_PM_DEVICE
```
ifdef ?

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,213 @@
+
+		res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	}
+end_trans:
```
in case you didn't locked the mutex successfully your are unlocking it there

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,213 @@
+		cb(dev, ret, context, arg);
+	}
+
+
```
remove blank line there

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,213 @@
+
+	if (!device_is_ready(config->bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus->name);
+		return -1;
```
maybe better to return an errno error

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,213 @@
+
+	if (!device_is_ready(cfg->root)) {
+		LOG_ERR("I2C mux root %s not ready", cfg->root->name);
+		return -1;
```
same here

kurddt commented:
> > Please let me know how to fix the CODEOWNER failure
> 
> You need to add entries in CODEOWNER file for the new files that this PR adds. Are you willing to maintain this driver (ie make fixes or changes as issues are raised)?

Yes I can maintain this driver if no one else want to take ownership

kurddt commented:
> > Please let me know how to fix the CODEOWNER failure
> 
> You need to add entries in CODEOWNER file for the new files that this PR adds. Are you willing to maintain this driver (ie make fixes or changes as issues are raised)?

It seems that most directories under `drivers/` have a default owner entry, any reason why it's not the case for `i2c` ?

mbolivar-nordic commented:
i2c is no longer maintained.

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,200 @@
+
+struct tca9546a_channel_config {
+	const struct device *root;
+	uint8_t addr;
```
Does it's an I2C address or the channel index ?

kurddt replied:
That's the channel index, I will rename it

julien-massot replied:
ok I see, in this case could be great to rename also tca9546a_set_selection to tca9546a_set_chan/tca9546a_set_channel, and select_reg to select_chan
see https://github.com/torvalds/linux/blob/master/drivers/i2c/muxes/i2c-mux-pca954x.c

kurddt replied:
Actually it's more of a mask inside the address register

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,200 @@
+	struct tca9546a_root_data *data = dev->data;
+	const struct tca9546a_root_config *cfg = dev->config;
+
+	if (data->select_reg != sel) {
```
looks like possible value for this chip is 0xF max may be great to have a check for bad input from DT.

kurddt commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,200 @@
+		return -ENODEV;
+	}
+
+	i2c_tca9546a->select_reg = 0;
```
Yes because `data->select_reg` is a bit field. So `tca9546a_set_selection` will never be called with '0' as parameter. So on the first transaction after initialization the transaction to set the hardware register will always happen.

julien-massot replied:
ok for me

julien-massot commented:
hi @kurddt I'm totally sorry I did delete one of your comment by mistake while doing my review. It was about checking channel parameter range on tca9546a_set_selection.

I will be curious to know if it's possible to restore a deleted comment.

kurddt commented:
> hi @kurddt I'm totally sorry I did delete one of your comment by mistake while doing my review. It was about checking channel parameter range on tca9546a_set_selection.

I'm actually trying to address it did you see my reply on that ?

kurddt commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,207 @@
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
```
Depending on some feedback on #35959 we might want to remove that

kurddt replied:
I removed the PM for now can always be merged later

julien-massot commented:
> > hi @kurddt I'm totally sorry I did delete one of your comment by mistake while doing my review. It was about checking channel parameter range on tca9546a_set_selection.
> 
> I'm actually trying to address it did you see my reply on that ?
Yes thanks for adding a channel validity verification.

julien-massot commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,168 @@
+
+	if (cfg->chan_mask > MAX_CHANNEL_MASK) {
+		LOG_ERR("Wrong DTS address provided for %s", dev->name);
+		return -EBADR;
```
I think -EINVAL is more appropriate but looks good anyway.

cfriedt commented:
@kurddt - please rebase and resolve conflicts.

Commit:
driver: i2c: Add TCA9546a I2C switch driver

The driver only support use case where the channels are used in mutual
exclusion.
Origin: Original

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

mbolivar-nordic commented on drivers/i2c/Kconfig.tca9546a:
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2020 Innoseis BV
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig I2C_TCA9546A
+	bool "I2C addressable switch"
+	help
+		Enable TCA9546A i2C bus switch
+
+if I2C_TCA9546A
+	config  I2C_TCA9546_ROOT_INIT_PRIO
+	int "TCA9546a root driver init priority"
+	default I2C_INIT_PRIORITY
+
+	config  I2C_TCA9546_CHANNEL_INIT_PRIO
+	int "TCA9546a channel driver init priority"
+	default I2C_INIT_PRIORITY
+endif
```
```suggestion
menuconfig I2C_TCA9546A
	bool "I2C addressable switch"
	help
	  Enable TCA9546A I2C bus switch

if I2C_TCA9546A

config I2C_TCA9546_ROOT_INIT_PRIO
	int "TCA9546a root driver init priority"
	default I2C_INIT_PRIORITY

config I2C_TCA9546_CHANNEL_INIT_PRIO
       int "TCA9546a channel driver init priority"
       default I2C_INIT_PRIORITY

endif # I2C_TCA9546A
```

The whitespace etc. here is very unusual to my eyes Please see above suggestion.

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,18 @@
+
+properties:
+    label:
+        required: true
```
Why?

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2020, Innoseis BV
+# SPDX-License-Identifier: Apache-2.0
+
+description: TCA9546a I2C Mux
```
I think it would be nice to give an example for how to put a sensor device underneath this thing here using a multiline comment in the description.

kurddt replied:
What do you mean by a multiline comment ?

kurddt replied:
LIke here https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twis.yaml#L4 ?

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,46 @@
+description: |
+  TCA9546a I2C switch
+
+  Each channel is represented by a separate device tree child node
```
```suggestion
  Each channel is represented by a separate devicetree child node
```

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,46 @@
+
+child-binding:
+  description: TCA9546a I2C switch channel node
+  compatible: "ti,tca9546a,channel"
```
Sorry, but I just noticed this. I think this should be `ti,tca9546a-channel` instead. Usually, `-` is used to separate words in the `device` portion of a `vnd,device` compatible.

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,46 @@
+  Each channel is represented by a separate device tree child node
+  with a "ti,tca9546a,channel" compatible.
+  The channel node can then be used as a standard i2c bus controller
+  like in the following simplified device tree:
```
```suggestion
  like in the following simplified example:
```

mbolivar-nordic reviewed (changes requested):
Yes, that's exactly what I meant by multiline comment. Thanks, this is very helpful. A couple of things I noticed while looking at the updated binding.

MaureenHelm commented on drivers/i2c/i2c_tca9546a.c:
```diff
@@ -0,0 +1,168 @@
+	return 0;
+}
+
+struct i2c_driver_api tca9546a_api_funcs = {
```
Add `const`

MaureenHelm reviewed (commented):
Looks good, but I'd like to make sure this builds in CI. Are there any in-tree boards with this device?

kurddt commented:
> Are there any in-tree boards with this device?

Not that I know of. Is there a place where to put a "dummy" devicetree in order to have the driver build but without running a test ?

mbolivar-nordic commented on drivers/i2c/Kconfig.tca9546a:
```diff
@@ -0,0 +1,19 @@
+menuconfig I2C_TCA9546A
+	bool "I2C addressable switch"
+	help
+		Enable TCA9546A i2C bus switch
```
```suggestion
	  Enable TCA9546A i2C bus switch
```

This one was not quite updated as suggested. Please keep it consistent with the rest of the kconfig help throughout the tree by using a tab and two spaces instead of two tabs. I know it's weird, but see e.g. Kconfig.zephyr; we inherit this from Linux and are generally consistent about it.

mbolivar-nordic commented on drivers/i2c/Kconfig.tca9546a:
```diff
@@ -0,0 +1,19 @@
+
+if I2C_TCA9546A
+
+config  I2C_TCA9546_ROOT_INIT_PRIO
```
```suggestion
config I2C_TCA9546_ROOT_INIT_PRIO
```

Being strict about whitespace consistency in this case makes it possible to search for the definition of the config option using `git grep 'config I2C_TCA9546_ROOT_INIT_PRIO'` instead of messing around with variable whitespace.

mbolivar-nordic commented on drivers/i2c/Kconfig.tca9546a:
```diff
@@ -0,0 +1,19 @@
+	int "TCA9546a root driver init priority"
+	default I2C_INIT_PRIORITY
+
+config  I2C_TCA9546_CHANNEL_INIT_PRIO
```
```suggestion
config I2C_TCA9546_CHANNEL_INIT_PRIO
```

mbolivar-nordic commented on dts/bindings/i2c/ti,tca9546a.yaml:
```diff
@@ -0,0 +1,46 @@
+  TCA9546a I2C switch
+
+  Each channel is represented by a separate devicetree child node
+  with a "ti,tca9546a,channel" compatible.
```
```suggestion
  with a "ti,tca9546a-channel" compatible.
```

mbolivar-nordic reviewed (changes requested):
Just a few more nitpicks from me.

MaureenHelm reviewed (commented):
Needs whitespace changes requested by @mbolivar-nordic , but otherwise LGTM. Thanks for adding a test.

mbolivar-nordic reviewed (approved):
My comments are addressed; thanks!
